### PR TITLE
Add e2e Integration tests for Reconcile loop

### PR DESCRIPTION
**Pull request by achouhan09**

### CODE REVIEW

File: tests/suite_test.go [hunk 1]
```diff
@@ -97,6 +109,15 @@ var _ = BeforeSuite(func() {
 		err = mgr.Start(ctrl.SetupSignalHandler())
 		Expect(err).ToNot(HaveOccurred())
 	}()
+
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: Namespace,
+		},
+	}
+	err = k8sClient.Create(context.Background(), namespace, &client.CreateOptions{})
```
Comment (naveenpaul1) [inline tests/suite_test.go:118]: delete the resources after the running the test suite

File: tests/managedmcg_controller_test.go [hunk 2]
```diff
@@ -50,3 +43,95 @@ var _ = Describe("ManagedMCG validations", func() {
...
+			newManagedMCG := managedMCG.DeepCopy()
+			err := k8sClient.Delete(context.TODO(), newManagedMCG, &client.DeleteOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			err = k8sClient.DeleteAllOf(context.TODO(), &corev1.Secret{}, &client.DeleteAllOfOptions{
+				ListOptions: client.ListOptions{
+					Namespace: Namespace,
+				},
+			})
+			Expect(err).NotTo(HaveOccurred())
```
Comment (rexagod) [inline tests/managedmcg_controller_test.go:112]: OCS CSV deletion also needs to happend here.

File: tests/managedmcg_controller_test.go [hunk 3]
```diff
@@ -50,3 +43,95 @@ var _ = Describe("ManagedMCG validations", func() {
...
+		})
+		It("should be able to read ManagedMCG object", func() {
```
Comment (rexagod) [inline tests/managedmcg_controller_test.go:114]:
```suggestion
		It("should be able to reconcile ManagedMCG object", func() {
```

File: tests/managedmcg_controller_test.go [hunk 4]
```diff
@@ -50,3 +43,95 @@ var _ = Describe("ManagedMCG validations", func() {
...
+			By("providing valid ManagedMCG name", func() {
```
Comment (rexagod) [inline tests/managedmcg_controller_test.go:115]:
```suggestion
			By("providing valid ManagedMCG params", func() {
```

File: tests/managedmcg_controller_test.go [hunk 5]
```diff
@@ -50,3 +43,95 @@ var _ = Describe("ManagedMCG validations", func() {
...
+				r.Client = fake.NewClientBuilder().WithScheme(r.Scheme).WithObjects(&ocscsv, &smtpSecretFake, &addonSecretFake, &deadMansSecretfake, &pagerDutySecretFake).Build()
+
+				req := ctrl.Request{
+					NamespacedName: types.NamespacedName{
+						Name:      "managedmcg",
+						Namespace: Namespace,
+					},
+				}
+
+				_, err := r.Reconcile(context.TODO(), req)
```
Comment (rexagod) [inline tests/managedmcg_controller_test.go:132]:
```suggestion
				_, err := r.Reconcile(context.Background(), req)
```

File: tests/managedmcg_controller_test.go [hunk 6]
```diff
@@ -50,3 +43,95 @@ var _ = Describe("ManagedMCG validations", func() {
...
+			err := k8sClient.Create(context.Background(), newManagedMCG, &client.CreateOptions{})
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		AfterEach(func() {
+			newManagedMCG := managedMCG.DeepCopy()
+			err := k8sClient.Delete(context.TODO(), newManagedMCG, &client.DeleteOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			err = k8sClient.DeleteAllOf(context.TODO(), &corev1.Secret{}, &client.DeleteAllOfOptions{
```
Comment (rexagod) [inline tests/managedmcg_controller_test.go:107]:
```suggestion
			err = k8sClient.DeleteAllOf(context.Background(), &corev1.Secret{}, &client.DeleteAllOfOptions{
```

File: tests/managedmcg_controller_test.go [hunk 7]
```diff
@@ -50,3 +43,95 @@ var _ = Describe("ManagedMCG validations", func() {
...
+
+		BeforeEach(func() {
+			newManagedMCG := managedMCG.DeepCopy()
+			err := k8sClient.Create(context.Background(), newManagedMCG, &client.CreateOptions{})
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		AfterEach(func() {
+			newManagedMCG := managedMCG.DeepCopy()
+			err := k8sClient.Delete(context.TODO(), newManagedMCG, &client.DeleteOptions{})
```
Comment (rexagod) [inline tests/managedmcg_controller_test.go:104]:
```suggestion
			err := k8sClient.Delete(context.Background(), newManagedMCG, &client.DeleteOptions{})
```

File: tests/managedmcg_controller_test.go [hunk 8]
```diff
@@ -50,3 +43,95 @@ var _ = Describe("ManagedMCG validations", func() {
...
+		It("should be able to read ManagedMCG object", func() {
+			By("providing valid ManagedMCG name", func() {
+
+				r := &controllers.ManagedMCGReconciler{
+					Scheme: k8sClient.Scheme(),
+				}
+
+				r.Log = ctrl.Log.WithName("controllers").WithName("ManagedMCGFake")
+
+				r.Client = fake.NewClientBuilder().WithScheme(r.Scheme).WithObjects(&ocscsv, &smtpSecretFake, &addonSecretFake, &deadMansSecretfake, &pagerDutySecretFake).Build()
```
Comment (rexagod) [inline tests/managedmcg_controller_test.go:123]:
```suggestion
				r := &controllers.ManagedMCGReconciler{
					Scheme: k8sClient.Scheme(),
					Log: ctrl.Log.WithName("controllers").WithName("ManagedMCGFake"),
					Client: fake.NewClientBuilder().WithScheme(r.Scheme).WithObjects(&ocscsv, &smtpSecretFake, &addonSecretFake, &deadMansSecretfake, &pagerDutySecretFake).Build()
				}
```

File: tests/suite_test.go (hunk 1 above)
Comment (achouhan09) [reply tests/suite_test.go:118]: Okay @naveen , I will add that for namespace as well. Thanks!

File: tests/suite_test.go [hunk 9]
```diff
@@ -33,13 +37,18 @@ import (
...
+	opv1a1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
 	mcgopenshiftiov1alpha1 "github.com/red-hat-storage/mcg-osd-deployer/api/v1alpha1"
 	//+kubebuilder:scaffold:imports
 )
 
 // These tests use Ginkgo (BDD-style Go testing framework). Refer to
 // http://onsi.github.io/ginkgo/ to learn more about Ginkgo.
 
+const (
+	Namespace = "redhat-data-federation"
```
Comment (rexagod) [inline tests/suite_test.go:49]:
```suggestion
	namespace = "redhat-data-federation"
```

No need to export the namespace. This usage will always be under the `tests` package.

**Review by rexagod — CHANGES_REQUESTED** [review]

File: tests/managedmcg_controller_test.go [hunk 10]
```diff
@@ -6,13 +6,15 @@ import (
...
 	. "github.com/onsi/gomega"
 	managedmcgv1alpha1 "github.com/red-hat-storage/mcg-osd-deployer/api/v1alpha1"
+	"github.com/red-hat-storage/mcg-osd-deployer/controllers"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-)
 
-const (
-	Namespace = "openshift-storage"
```
Comment (rexagod) [inline tests/managedmcg_controller_test.go:15]: Let's keep the `namespace` here itself. `suite_test.go` is meant for setting up extraneous setup logic for the fake cluster, and it's better to couple similar logic under the same test file.

**Review by rexagod — CHANGES_REQUESTED** [review]

File: tests/suite_test.go (hunk 1 above)
Comment (rexagod) [reply tests/suite_test.go:118]: PTAL: https://github.com/red-hat-storage/mcg-osd-deployer/pull/29#discussion_r866836266

`env-test` will eventually delete all the remaining resources during the tear down of the fake cluster we test e2e features in, but I agree with Naveen that it's a good practice to delete resources as long as they are not needed anymore, just so we are sure that the flow continues in the the absence of them, and is as expected.

File: tests/managedmcg_controller_test.go (hunk 10 above)
Comment (achouhan09) [reply tests/managedmcg_controller_test.go:15]: Okay.

Comment (achouhan09) [reply tests/managedmcg_controller_test.go:15]: @rexagod , Here is an issue while creating and deleting separate namespace for both specs. As both specs are executing in parallel and we are trying to accessing the ns from one spec which was already in terminating state by other spec.

Comment (rexagod) [reply tests/managedmcg_controller_test.go:15]: Parallelization only happens if we pass the `-p` flag to Ginkgo, which is not the case right now. Are you sure about this?

Comment (rexagod) [reply tests/managedmcg_controller_test.go:15]: Tests should themselves be mutually exclusive in execution, and any race conditions should be avoided while writing them. A good practice is to imagine the the specs will execute in parallel, even if that's not actually the case, and write your spec based on that logic. In this case NS creation and deletion would happen for both specs, if they require it to exist.

File: tests/managedmcg_controller_test.go [hunk 11]
```diff
@@ -89,47 +113,64 @@ var _ = Describe("ManagedMCGReconciler Reconcile", func() {
...
+
+			// err = fake.NewClientBuilder().WithScheme(k8sClient.Scheme()).Build().Delete(context.Background(), &noobaacsv, &client.DeleteOptions{})
+			// Expect(err).NotTo(HaveOccurred())
+
+			// err = k8sClient.Delete(context.Background(), ns, &client.DeleteOptions{})
+			// Expect(err).NotTo(HaveOccurred())
 		})
-		It("should be able to read ManagedMCG object", func() {
-			By("providing valid ManagedMCG name", func() {
+		It("should be able to reconclie ManagedMCG object", func() {
```
Comment (rexagod) [inline tests/managedmcg_controller_test.go:157]: `s/reconclie/reconcile`

**Review by rexagod — CHANGES_REQUESTED** [review]
Please remove all the commented code.

File: tests/managedmcg_controller_test.go [hunk 12]
```diff
@@ -6,45 +6,189 @@ import (
...
+		newNamespace := ns.DeepCopy()
+		newManagedMCG := managedMCG.DeepCopy()
+
+		addonSecretFake := corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "addon-secret-fake",
+				Namespace: namespace,
+			},
+		}
+		addonSecretFake.Data = make(map[string][]byte, 1)
```
Comment (naveenpaul1) [inline tests/managedmcg_controller_test.go:86]: just like ObjectMeta can you add Data also inside the struct

Comment (achouhan09) [reply tests/managedmcg_controller_test.go:86]: okay @naveenpaul1 , I will do that. Thanks!

File: tests/managedmcg_controller_test.go [hunk 13]
```diff
@@ -6,45 +6,189 @@ import (
...
+				Expect(err).NotTo(HaveOccurred())
+
+				err = r.Client.Delete(context.Background(), newManagedMCG, &client.DeleteOptions{})
+				Expect(err).NotTo(HaveOccurred())
+
+				err = r.Client.Get(context.Background(), client.ObjectKeyFromObject(newNamespace), newNamespace)
+				Expect(err).NotTo(HaveOccurred())
+
+				err = r.Client.Delete(context.Background(), newNamespace, &client.DeleteOptions{})
 				Expect(err).NotTo(HaveOccurred())
```
Comment (rexagod) [inline tests/managedmcg_controller_test.go:69]: Why are we testing namespace creation and deletion **after** `ManagedMCG` has already been created and deleted? IMO the flow should be:

1. NS is created.
2. MMCG is created in that NS.
3. MMCG is deleted.
4. NS is deleted.

File: tests/managedmcg_controller_test.go [hunk 14]
```diff
@@ -6,45 +6,189 @@ import (
...
+		addonSecretFake.Data["addonparam"] = []byte("foo")
+
+		pagerDutySecretFake := corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "pagerduty-secret-fake",
+				Namespace: namespace,
+			},
+		}
+		pagerDutySecretFake.Data = make(map[string][]byte, 1)
+		pagerDutySecretFake.Data["PAGERDUTY_KEY"] = []byte("foo")
```
Comment (rexagod) [inline tests/managedmcg_controller_test.go:96]: https://github.com/red-hat-storage/mcg-osd-deployer/pull/29/files#r873624169

File: tests/managedmcg_controller_test.go [hunk 15]
```diff
@@ -6,45 +6,189 @@ import (
...
+
+		deadMansSecretfake := corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "deadmans-secret-fake",
+				Namespace: namespace,
+			},
+		}
+		deadMansSecretfake.Data = make(map[string][]byte, 1)
+		deadMansSecretfake.Data["SNITCH_URL"] = []byte("foo")
```
Comment (rexagod) [inline tests/managedmcg_controller_test.go:105]: https://github.com/red-hat-storage/mcg-osd-deployer/pull/29/files#r873624169

File: tests/managedmcg_controller_test.go [hunk 16]
```diff
@@ -6,45 +6,189 @@ import (
...
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "smtp-secret-fake",
+				Namespace: namespace,
+			},
+		}
+		smtpSecretFake.Data = make(map[string][]byte, 1)
+		smtpSecretFake.Data["host"] = []byte("host")
+		smtpSecretFake.Data["port"] = []byte("8080")
+		smtpSecretFake.Data["username"] = []byte("username")
+		smtpSecretFake.Data["password"] = []byte("password")
```
Comment (rexagod) [inline tests/managedmcg_controller_test.go:117]: https://github.com/red-hat-storage/mcg-osd-deployer/pull/29/files#r873624169

File: tests/managedmcg_controller_test.go [hunk 17]
```diff
@@ -6,45 +6,189 @@ import (
...
+		noobacsv := noobaav1alpha1.NooBaa{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "nooba-operator",
+				Namespace: namespace,
+			},
+		}
+
+		BeforeEach(func() {
+			r.Scheme = k8sClient.Scheme()
+			r.Client = fake.NewClientBuilder().WithScheme(k8sClient.Scheme()).WithObjects(newNamespace, newManagedMCG, &ocscsv, &smtpSecretFake, &deadMansSecretfake, &pagerDutySecretFake, &noobacsv).Build()
```
Comment (rexagod) [inline tests/managedmcg_controller_test.go:135]: We don't need all of these resource for **each** test case. Please put only those resource here that are common in all cases, and create the uncommon ones inside their specific cases. This helps make cases more predictable, and sensible in general.

File: tests/managedmcg_controller_test.go [hunk 18]
```diff
@@ -6,45 +6,189 @@ import (
...
+			Expect(err).NotTo(HaveOccurred())
+
+			err = r.Client.Delete(context.Background(), newManagedMCG, &client.DeleteOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			err = r.Client.Get(context.Background(), client.ObjectKeyFromObject(newNamespace), newNamespace)
+			Expect(err).NotTo(HaveOccurred())
+
+			err = r.Client.Delete(context.Background(), newNamespace, &client.DeleteOptions{})
+			Expect(err).NotTo(HaveOccurred())
```
Comment (rexagod) [inline tests/managedmcg_controller_test.go:179]: Same as `BeforeEach` comment above, but for deletion.

**Review by rexagod — CHANGES_REQUESTED** [review]

File: tests/managedmcg_controller_test.go (hunk 13 above)
Comment (achouhan09) [reply tests/managedmcg_controller_test.go:69]: We are checking if ns is already present or not before deletion.
Okay , we can remove that before check.

File: tests/managedmcg_controller_test.go (hunk 18 above)
Comment (achouhan09) [reply tests/managedmcg_controller_test.go:179]: yes ,I will remove the before check from here as well. Thanks

**Comment by achouhan09** [comment]
@naveenpaul1 @rexagod Please let me know if it need any more changes or it is ready to merge.

**Pull request merged**